README.md
# next-pre-compression

[![NPM version][npm-image]][npm-url] [![Downloads][downloads-image]][npm-url] [![Build Status][travis-image]][travis-url] [![Coverage Status][codecov-image]][codecov-url] [![Dependency status][david-dm-image]][david-dm-url] [![Dev Dependency status][david-dm-dev-image]][david-dm-dev-url]

[npm-url]:https://npmjs.org/package/@moxy/next-pre-compression
[downloads-image]:https://img.shields.io/npm/dm/@moxy/next-pre-compression.svg
[npm-image]:https://img.shields.io/npm/v/@moxy/next-pre-compression.svg
[travis-url]:https://travis-ci.org/moxystudio/next-pre-compression
[travis-image]:http://img.shields.io/travis/moxystudio/next-pre-compression/master.svg
[codecov-url]:https://codecov.io/gh/moxystudio/next-pre-compression
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/next-pre-compression/master.svg
[david-dm-url]:https://david-dm.org/moxystudio/next-pre-compression
[david-dm-image]:https://img.shields.io/david/moxystudio/next-pre-compression.svg
[david-dm-dev-url]:https://david-dm.org/moxystudio/next-pre-compression?type=dev
[david-dm-dev-image]:https://img.shields.io/david/dev/moxystudio/next-pre-compression.svg

Next.js plugin to compress static assets at build time and serve them instead of having to compress on-the-fly.

> ℹ️ While compressing ahead of time has benefits, it's preferrable to delegate compression to an upstream server, such as a reverse-proxy CDN. Modern CDNs apply compression with the best algorithms and settings, while storing the compressed responses on edge servers. Next.js already does `gzip` by default which is sufficient for efficient transfer of data to the upstream servers. In conclusion, you should only use this package if there's no upstream server or if it doesn't apply compression.

## Installation

```sh
$ npm i --save @moxy/next-pre-compression
```

## Usage

### next.config.js

Setup the plugin in the `next.config.js` file:

```js
const withPreCompression = require('@moxy/next-pre-compression');

module.exports = withPreCompression({ ...nextConfig });
```

This plugin will automatically disable itself if you disable [`compress`](https://nextjs.org/docs#compression) in your `next.config.js`.

> ℹ️ The plugin has no effect in development.

### Express

Express middleware used to serve the previously compressed files, by leveraging [express-static-gzip](https://www.npmjs.com/package/express-static-gzip).

First, you need to setup a [custom express server]( https://github.com/zeit/next.js/tree/master/examples/custom-server-express). Then, simply add the middleware like so:

```js
// server.js

const express = require('express');
const next = require('next');
const preCompression = require('@moxy/next-pre-compression/express-middleware');
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
    const server = express();

    server.use(preCompression(app, {
        maxAge: 30 * 24 * 60 * 60 * 1000, // 30 days
    }));

    server.get('*', (req, res) => handle(req, res));

    server.listen(port, host, (err) => {
        if (err) { throw err; }

        console.log(`> Ready on http://localhost:${port}`);
    });
})
.catch((err) => {
    setImmediate(() => { throw err; });
});
```

> ℹ️ The middleware has no effect in development.

> ⚠️ A custom `assetPrefix` that references an absolute URI is not yet supported (e.g.: https://cdn.my-site.com), see [moxystudio/next-pre-compression#8](https://github.com/moxystudio/next-pre-compression/issues/8).

All options from [serve-static](https://www.npmjs.com/package/serve-static) are also available.

> ⚠️ You can't enable the `index` option as it's always set to false, due to a strange behavior of `express-static-gzip` that modifies the request path without restoring it, see: https://github.com/tkoenig89/express-static-gzip/blob/94767f79e861a3901e8ebba31b084abc4986817f/index.js#L28

##### Default options

```js
{
    maxAge: 365 * 24 * 60 * 60 * 1000, // 1 year
    immutable: true,
    etag: false,
    index: false // Can't be changed
}
```

## Tests

Any parameter passed to the `test` command, is passed down to Jest.

```sh
$ npm test
$ npm test -- --watch # during development
```

After running the tests, a **coverage** folder will be created containing the test coverage info.

## License

Released under the [MIT License](http://www.opensource.org/licenses/mit-license.php).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.0.1](https://github.com/moxystudio/next-pre-compression-plugin/compare/v3.0.0...v3.0.1) (2019-10-23)


### Bug Fixes

* disable etag ([c0d5e28](https://github.com/moxystudio/next-pre-compression-plugin/commit/c0d5e28))

## [3.0.0](https://github.com/moxystudio/next-pre-compression-plugin/compare/v2.1.0...v3.0.0) (2019-10-22)


### ⚠ BREAKING CHANGES

* changed the name of the package
* changed interface to pass next's app to middleware
* changed options to only allow serve-static options

* package usage ([6ca46c7](https://github.com/moxystudio/next-pre-compression-plugin/commit/6ca46c7))

## [2.1.0](https://github.com/moxystudio/next-pre-compression-plugin/compare/v2.0.2...v2.1.0) (2019-10-22)


### Features

* respect compress in next.config.js ([be6cf21](https://github.com/moxystudio/next-pre-compression-plugin/commit/be6cf21))

### [2.0.2](https://github.com/moxystudio/next-pre-compression-plugin/compare/v2.0.1...v2.0.2) (2019-10-21)


### Bug Fixes

* do not compress on the server compilation ([40303fe](https://github.com/moxystudio/next-pre-compression-plugin/commit/40303fe))

### [2.0.1](https://github.com/moxystudio/next-pre-compression-plugin/compare/v2.0.0...v2.0.1) (2019-10-10)


### Bug Fixes

* fix next peer dependency range ([0f69fc1](https://github.com/moxystudio/next-pre-compression-plugin/commit/0f69fc1))

## [2.0.0](https://github.com/moxystudio/next-pre-compression-plugin/compare/v1.0.2...v2.0.0) (2019-10-10)


### ⚠ BREAKING CHANGES

* the way the next plugin is imported has changed

### Bug Fixes

* export defaults to plugin ([#6](https://github.com/moxystudio/next-pre-compression-plugin/issues/6)) ([e6bafaf](https://github.com/moxystudio/next-pre-compression-plugin/commit/e6bafaf))

### [1.0.2](https://github.com/moxystudio/next-pre-compression-plugin/compare/v1.0.1...v1.0.2) (2019-09-30)


### Bug Fixes

* middleware naming fix ([41a57dd](https://github.com/moxystudio/next-pre-compression-plugin/commit/41a57dd))

### [1.0.1](https://github.com/moxystudio/next-pre-compression-plugin/compare/v1.0.0...v1.0.1) (2019-09-27)

## 1.0.0 (2019-09-27)


### Features

* initial implementation ([c0479ef](https://github.com/moxystudio/next-pre-compression-plugin/commit/c0479ef))


test/plugin.test.js
'use strict';

const zlib = require('zlib');
const withPreCompression = require('..');

const zlibBrotliCompression = zlib.brotliCompress;

const nextConfig = {
    compress: true,
};

let config = {
    plugins: [],
};

beforeEach(() => {
    config = {
        plugins: [],
    };
});

afterEach(() => {
    zlib.brotliCompress = zlibBrotliCompression;
});

it('should return the same config object that is passed', () => {
    const result = withPreCompression(nextConfig).webpack(config, {});

    expect(result).toBe(config);
});

it('should push compression settings to webpack config', () => {
    withPreCompression(nextConfig).webpack(config, {});

    expect(config).toMatchSnapshot();
});

it('should do nothing if compilation is associated to the server', () => {
    withPreCompression(nextConfig).webpack(config, { isServer: true });

    expect(config.plugins).toHaveLength(0);
});

it('should do nothing if compress is disable in the next config', () => {
    withPreCompression({ ...nextConfig, compress: false }).webpack(config, {});

    expect(config.plugins).toHaveLength(0);
});

it('should call nextConfig webpack if defined', () => {
    const nextConfig = {
        webpack: jest.fn(() => 'foo'),
    };

    const result = withPreCompression(nextConfig).webpack(config, {});

    expect(nextConfig.webpack).toHaveBeenCalledTimes(1);
    expect(nextConfig.webpack).toHaveReturnedWith(result);
});

it('should not push brotli compression plugin if brotli compression is not available', () => {
    delete zlib.brotliCompress;

    withPreCompression(nextConfig).webpack(config, {});

    expect(config).toMatchSnapshot();
});


test/express-middleware.test.js
'use strict';

const zlib = require('zlib');
const path = require('path');
const express = require('express');
const request = require('supertest');
const preCompressionMiddleware = require('../express-middleware');

const NEXT_STATIC_PATH = '/_next/static/';

const NEXT_APP = {
    nextConfig: {
        distDir: path.join(__dirname, 'fixtures'),
        assetPrefix: '',
    },
    renderOpts: {
        dev: false,
    },
};

it('should respond success if requested from static resources (gzip)', async () => {
    const server = express();

    server.use(preCompressionMiddleware(NEXT_APP));

    await request(server)
    .get(`${NEXT_STATIC_PATH}test.txt`)
    .set('Accept-Encoding', 'gzip')
    .expect(200, 'hello world\n')
    .expect('Content-Encoding', 'gzip');
});

it('should respond success if requested from static resources (brotli)', async () => {
    const server = express();

    server.use(preCompressionMiddleware(NEXT_APP));

    await request(server)
    .get(`${NEXT_STATIC_PATH}test.txt`)
    .set('Accept-Encoding', 'br')
    .expect(200)
    .expect('Content-Encoding', 'br')
    .expect((res) => {
        // We're decompressing manually since SuperAgent doesn't support Brotli at the moment
        const decompressedPayload = zlib.brotliDecompressSync(Buffer.from(res.text));

        return decompressedPayload === 'hello world\n';
    });
});

it('should call response with 404 if request file doesn\'t exist', async () => {
    const server = express();

    server.use(preCompressionMiddleware(NEXT_APP));

    await request(server)
    .get(`${NEXT_STATIC_PATH}somefilethatwillneverexist.txt`)
    .expect(404);
});

it('should work correctly with a custom assetPrefix', async () => {
    const server = express();

    server.use(preCompressionMiddleware({
        ...NEXT_APP,
        nextConfig: {
            ...NEXT_APP.nextConfig,
            assetPrefix: '/foo',
        },
    }));

    await request(server)
    .get(`/foo${NEXT_STATIC_PATH}/test.txt`)
    .set('Accept-Encoding', 'gzip')
    .expect(200, 'hello world\n');
});

it('should throw if a custom assetPrefix is an absolute URI', () => {
    const server = express();

    expect(() => {
        server.use(preCompressionMiddleware({
            ...NEXT_APP,
            nextConfig: {
                ...NEXT_APP.nextConfig,
                assetPrefix: 'https://cdn.my-site.com',
            },
        }));
    }).toThrow(/doesn't support absolute URI/);
});

it('should do nothing in dev', async () => {
    const server = express();

    server.use(preCompressionMiddleware({
        ...NEXT_APP,
        renderOpts: {
            ...NEXT_APP.renderOpts,
            dev: true,
        },
    }));

    server.get('*', (req, res) => res.send('foo'));

    await request(server)
    .get(`${NEXT_STATIC_PATH}test.txt`)
    .set('Accept-Encoding', 'gzip')
    .expect(200, 'foo');
});

it('should respond with the correct default headers', async () => {
    const server = express();

    server.use(preCompressionMiddleware(NEXT_APP));

    await request(server)
    .get(`${NEXT_STATIC_PATH}/test.txt`)
    .expect('Cache-Control', 'public, max-age=31536000, immutable')
    .expect((res) => {
        expect(res.headers.etag).toBe(undefined);
    });
});

it('should forward options to serve-static', async () => {
    const server = express();

    server.use(preCompressionMiddleware(NEXT_APP, {
        maxAge: 60 * 60 * 1000,
        immutable: false,
    }));

    await request(server)
    .get(`${NEXT_STATIC_PATH}/test.txt`)
    .expect('Cache-Control', 'public, max-age=3600');
});


test/express-compression.test.js
'use strict';

const express = require('express');
const request = require('supertest');
const compressionMiddleware = require('../express-compression');
let server = express();

beforeEach(() => {
    server = express();
});

it('should call next() if request path does not match', async () => {
    server.use(compressionMiddleware({
        requestPath: '/',
    }));

    server.get('/request/path', (req, res) => {
        res.status(200).send('foo');
    });

    await request(server)
    .get('/request/path')
    .expect(200, 'foo');
});

it('should call next() if request path matches and file does not exist', async () => {
    server.use(compressionMiddleware({
        requestPath: '/',
        fsPath: '/test_dir/nonexistent/directory',
    }));

    server.get('/', (req, res) => {
        res.status(200).send('foo');
    });

    await request(server)
    .get('/')
    .expect(200, 'foo');
});

it('should call next() with an error when staticGzipMiddleware fails', async () => {
    server.use(compressionMiddleware({
        requestPath: '/',
    }));

    server.get('/', (req, res) => {
        res.status(200).send('foo');
    });

    await request(server)
    .get('/')
    .expect(200, 'foo');
});

it('should set index path to index.html if none is set', async () => {
    server.use(compressionMiddleware({
        requestPath: '/',
    }));

    server.get('/', (req, res) => {
        res.status(200).send('foo');
    });

    await request(server)
    .get('/')
    .expect(200, 'foo');
});

it('should use index path string if none is set', async () => {
    server.use(compressionMiddleware({
        requestPath: '/',
        index: '/main.js',
    }));

    server.get('/', (req, res) => {
        res.status(200).send('foo');
    });

    await request(server)
    .get('/')
    .expect(200, 'foo');
});

it('should use index path string if index is false', async () => {
    server.use(compressionMiddleware({
        requestPath: '/',
        index: false,
    }));

    server.get('/', (req, res) => {
        res.status(200).send('foo');
    });

    await request(server)
    .get('/')
    .expect(200, 'foo');
});


index.js
/* eslint-disable no-useless-escape */

'use strict';

const CompressionPlugin = require('compression-webpack-plugin');
const zlib = require('zlib');
const mimeDb = require('mime-db');

module.exports = (nextConfig = {}) => ({
    ...nextConfig,
    webpack(config, options) {
        const enabled = nextConfig.compress && !options.isServer && !options.dev;

        if (enabled) {
            const compressibleRegExps = Object
            .values(mimeDb)
            .filter((mime) => mime.compressible && mime.extensions)
            .reduce((extensions, mime) => {
                mime.extensions.forEach((ext) => extensions.push(new RegExp(`\\.${ext}`)));

                return extensions;
            }, []);

            const compressionOptions = {
                test: /^static[\\\/]/,
                include: compressibleRegExps,
                exclude: /^static[\\\/]webpack[\\\/]/,
            };

            config.plugins.push(new CompressionPlugin({
                ...compressionOptions,
                filename: '[path].gz[query]',
                algorithm: 'gzip',
            }));

            if (zlib.brotliCompress) {
                config.plugins.push(new CompressionPlugin({
                    ...compressionOptions,
                    filename: '[path].br[query]',
                    algorithm: 'brotliCompress',
                }));
            }
        }

        if (typeof nextConfig.webpack === 'function') {
            return nextConfig.webpack(config, options);
        }

        return config;
    },
});


express-compression.js
'use strict';

const path = require('path');
const expressStaticGzip = require('express-static-gzip');
const { Router: createRouter } = require('express');

// https://github.com/tkoenig89/express-static-gzip/blob/94767f79e861a3901e8ebba31b084abc4986817f/index.js#L28
// expressStaticGzip has a function that changes the empty directory url to /index.html
// this behaviour is configurable through the expressStaticGzip option 'index'.
// maybeRestoreUrlFromIndexHtmlToEmpty restores the behaviour previously described
const maybeRestoreUrlFromIndexHtmlToEmpty = (req, options) => {
    if (options.index === false || !req.originalUrl.endsWith('/')) {
        return;
    }

    const indexStr = options.index == null ? 'index.html' : options.index;

    req.url = req.url.slice(0, -indexStr.length);
};

const compressionMiddleware = (options) => {
    options = {
        fsPath: '.next/static/',
        requestPath: '/_next/static/',
        orderPreference: ['br', 'gzip'],
        enableBrotli: true,
        serveStatic: {
            maxAge: 365 * 24 * 60 * 60 * 1000, // 1 year
            immutable: true,
        },
        ...options,
    };

    const router = createRouter();
    const { fsPath, requestPath, ...staticGzipOptions } = options;
    const staticGzipMiddleware = expressStaticGzip(path.resolve(fsPath), staticGzipOptions);

    return router.use(requestPath, (req, res, next) => {
        staticGzipMiddleware(req, res, (err) => {
            /* istanbul ignore if */
            if (err) {
                return next(err);
            }

            maybeRestoreUrlFromIndexHtmlToEmpty(req, options);

            next();
        });
    });
};

module.exports = compressionMiddleware;


next-webpack-compression.js
'use strict';

const CompressionPlugin = require('compression-webpack-plugin');
const zlib = require('zlib');
const mimeDb = require('mime-db');

module.exports = (nextConfig = {}) => ({
    ...nextConfig,
    webpack(config, options) {
        if (nextConfig.compress && !options.isServer) {
            const compressibleRegExps = Object
            .values(mimeDb)
            .filter((mime) => mime.compressible && mime.extensions)
            .reduce((extensions, mime) => {
                mime.extensions.forEach((ext) => extensions.push(new RegExp(`\\.${ext}`)));

                return extensions;
            }, []);

            config.plugins.push(new CompressionPlugin({
                filename: '[path].gz[query]',
                algorithm: 'gzip',
                include: compressibleRegExps,
            }));

            if (zlib.brotliCompress) {
                config.plugins.push(new CompressionPlugin({
                    filename: '[path].br[query]',
                    algorithm: 'brotliCompress',
                    include: compressibleRegExps,
                }));
            }
        }

        if (typeof nextConfig.webpack === 'function') {
            return nextConfig.webpack(config, options);
        }

        return config;
    },
});


express-middleware.js
'use strict';

const path = require('path');
const expressStaticGzip = require('express-static-gzip');
const { Router: createRouter } = require('express');

const NEXT_STATIC_PATH = '/_next/static/';

/**
 * Express middleware to serve pre-compressed files from filesystem.
 *
 * @param {object} app - Next.js' application.
 * @param {object} options - Options that will be passed to 'serve-static'.
 * @returns {Function} The middleware function.
 */
const preCompression = (app, options = {}) => {
    const { nextConfig: { distDir, assetPrefix }, renderOpts } = app;

    if (renderOpts.dev) {
        return (req, res, next) => next();
    }

    options = {
        maxAge: 365 * 24 * 60 * 60 * 1000, // 1 year
        immutable: true,
        etag: false,
        ...options,
        index: false,
    };

    const staticGzipMiddleware = expressStaticGzip(path.join(distDir, 'static'), {
        enableBrotli: true,
        orderPreference: ['br', 'gzip'],
        index: false,
        serveStatic: options,
    });

    if (/^[^:]+:\/\//.test(assetPrefix)) {
        throw new Error('The @moxy/next-pre-compression\'s express middleware doesn\'t support absolute URI as the assetPrefix yet (e.g.: https://cdn.my-site.com), see https://github.com/moxystudio/next-pre-compression/issues/8.');
    }

    const router = createRouter();
    const routePath = `${assetPrefix}${NEXT_STATIC_PATH}`;

    return router.use(routePath, staticGzipMiddleware);
};

module.exports = preCompression;
